lxd/storage/drivers/driver_dir.go

```diff
@@ -0,0 +1,864 @@
+	if err != nil {
+		if ourMount {
+			vol.Unmount(op)
+		}
+
+		return err
+	}
+
+	if ourMount {
+		vol.Unmount(op)
+	}
+
+	return nil
+}
+
+// CreateVolumeFromMigration creates a volume being sent via a migration.
+func (d *dir) CreateVolumeFromMigration(vol Volume, conn *websocket.Conn, volTargetArgs migration.VolumeTargetArgs, op *operations.Operation) error {
+	if vol.ContentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	if volTargetArgs.MigrationType.FSType != migration.MigrationFSType_RSYNC {
+		return fmt.Errorf("Migration type not supported")
+
+	}
+
+	volPath := vol.MountPoint()
+
+	// Get the volume ID for the new volumes, which is used to set project quota.
+	volID, err := d.getVolID(vol.Type, vol.Name)
+	if err != nil {
+		return err
+	}
+
+	// Create slice of paths created if revert needed later.
+	revertPaths := []string{}
+	defer func() {
+		// Remove any paths created if we are reverting.
+		for _, path := range revertPaths {
+			d.deleteQuota(path, volID)
+			os.RemoveAll(path)
+		}
+	}()
+
+	// Snapshots are sent first by the sender, so create these first.
+	for _, snapName := range volTargetArgs.Snapshots {
+		fullSnapName := fmt.Sprintf("%s%s%s", vol.Name, shared.SnapshotDelimiter, snapName)
+		snapshot := NewVolume(d, d.name, vol.Type, vol.ContentType, fullSnapName, vol.Config)
+		snapPath := snapshot.MountPoint()
+
+		err := os.MkdirAll(snapPath, 0711)
+		if err != nil {
+			return err
+		}
+
+		revertPaths = append(revertPaths, snapPath)
+
+		// Initialise the snapshot's quota with the parent volume's ID.
+		err = d.initQuota(snapPath, volID)
+		if err != nil {
+			return err
+		}
+
+		ourMount, err := snapshot.Mount(op)
+		if err != nil {
+			return err
+		}
+
+		wrapper := migration.ProgressWriter(op, "fs_progress", fullSnapName)
+		path := shared.AddSlash(snapshot.MountPoint())
+
+		err = rsync.Recv(path, conn, wrapper, volTargetArgs.MigrationType.Features)
+		if err != nil {
+			if ourMount {
+				snapshot.Unmount(op)
+			}
+
+			return err
+		}
+
+		if ourMount {
+			snapshot.Unmount(op)
+		}
+	}
+
+	// Finally the actual volume is sent by sender, so create that last.
+	err = os.MkdirAll(volPath, 0711)
+	if err != nil {
+		return err
+	}
+
+	revertPaths = append(revertPaths, volPath)
+
+	// Initialise the volume's quota using the volume ID.
+	err = d.initQuota(volPath, volID)
+	if err != nil {
+		return err
+	}
+
+	// Set the quota if specified in volConfig or pool config.
+	err = d.setQuota(volPath, volID, vol.Config["size"])
+	if err != nil {
+		return err
+	}
+
+	ourMount, err := vol.Mount(op)
+	if err != nil {
+		return err
+	}
+
+	wrapper := migration.ProgressWriter(op, "fs_progress", vol.Name)
+	path := shared.AddSlash(vol.MountPoint())
+	err = rsync.Recv(path, conn, wrapper, volTargetArgs.MigrationType.Features)
+	if err != nil {
+		if ourMount {
+			vol.Unmount(op)
```

Comment:
instead of having this in the normal and error path, you could just do right after the Mount call:
```go
if ourMount {
    defer vol.Unmount(op)
}
```

Reply:
Actually, scrap that, we're inside the `dir` backend, we know that `MountVolume()` and `UnmountVolume()` are no-ops, so we don't even need to jump through all those hoops :)

Reply:
So all the Mount and Unmount calls inside this function can go away.

Reply:
@stgraber yeah agreed, my main reason for adding these unnecessary noop mount/unmount calls was to get it clear in my mind where these would need to occur in future driver types that do need mounts and that my design had accommodated for them. Happy to remove them now though.

Reply:
@stgraber I thought a bit more about this, both taking into your point about removing the unnecessary mount code in dir whilst still keeping the dir implementation indicative of how other drivers should do it in the future. I've come up with adding a function to `func (v Volume) MountTask(task func(mountPath string, op *operations.Operation) error, op *operations.Operation) error` that takes a function to execute (the action you want to perform on the mounted volume) and then takes care of calling the driver's `Mount` and `Unmount` as needed (taking into account the `ourMount` response from `Mount`).

This feels like a good balance, as it reduces the mounting code in `dir`, and yet will mean that `dir` can behave similarly to other drivers that do actually need to mount.

Also, as Volume's `Mount` and `Unmount` are no longer actually used, I've removed them until they are needed.

Reply:
That looks pretty clean indeed and should avoid bad uses of mount/unmount by new implementations.